storage/imx6ull-flash/flashdev.c

```diff
@@ -0,0 +1,434 @@
+		err = _flashmtd_errRead(res, strg->dev->ctx->metabuf, err, strg->dev->ctx->databuf);
+		if (err < 0) {
+			break;
+		}
+
+		chunksz = ((len - tempsz) > strg->dev->mtd->writesz) ? strg->dev->mtd->writesz : (len - tempsz);
+		memcpy(data + tempsz, strg->dev->ctx->databuf, chunksz);
+		tempsz += chunksz;
+		offs += chunksz;
+	}
+
+	mutexUnlock(strg->dev->ctx->lock);
+
+	return tempsz > 0 ? tempsz : err;
+}
+
+
+static ssize_t flashmtd_write(struct _storage_t *strg, off_t offs, const void *data, size_t len)
+{
+	ssize_t res = 0;
+	blkcnt_t pageID;
+	size_t tempsz = 0;
+
+	if ((offs % strg->dev->mtd->writesz) != 0 || (len % strg->dev->mtd->writesz) != 0) {
+		return -EINVAL;
+	}
+
+	mutexLock(strg->dev->ctx->lock);
+	while (tempsz < len) {
+		/* TODO: should we skip badblocks ? */
+		memcpy(strg->dev->ctx->databuf, data + tempsz, strg->dev->mtd->writesz);
+
+		pageID = offs / strg->dev->mtd->writesz;
+		res = flashdrv_write(strg->dev->ctx->dma, pageID, strg->dev->ctx->databuf, NULL);
+		if (res < 0) {
+			break;
+		}
+
+		tempsz += strg->dev->mtd->writesz;
+		offs += strg->dev->mtd->writesz;
+	}
+	mutexUnlock(strg->dev->ctx->lock);
+
+	return tempsz > 0 ? tempsz : res;
+}
+
+
+static ssize_t flashmtd_metaRead(struct _storage_t *strg, off_t offs, void *data, size_t len)
+{
+	int res = 0, err = 0;
+	size_t tempsz = 0, chunksz;
+	flashdrv_meta_t *meta = strg->dev->ctx->metabuf;
+
+	if ((offs % strg->dev->mtd->writesz) != 0) {
+		return -EINVAL;
+	}
+
+	mutexLock(strg->dev->ctx->lock);
+	while (tempsz < len) {
+		memset(meta->errors, 0, sizeof(meta->errors));
+		/* TODO: should we skip badblocks ? */
+		res = flashdrv_read(strg->dev->ctx->dma, offs / strg->dev->mtd->writesz, NULL, strg->dev->ctx->metabuf);
+		err = _flashmtd_errRead(res, strg->dev->ctx->metabuf, err, strg->dev->ctx->databuf);
+		if (err < 0) {
+			break;
+		}
+
+		chunksz = len > strg->dev->mtd->oobSize ? strg->dev->mtd->oobSize : len;
+		memcpy(data + tempsz, strg->dev->ctx->metabuf, chunksz);
+
+		tempsz += chunksz;
+		offs += strg->dev->mtd->writesz;
+	}
+	mutexUnlock(strg->dev->ctx->lock);
+
+	return tempsz > 0 ? tempsz : err;
+}
+
+
+static ssize_t flashmtd_metaWrite(struct _storage_t *strg, off_t offs, const void *data, size_t len)
+{
+	int res;
+	size_t tempsz = 0, chunksz;
+
+	if ((offs % strg->dev->mtd->writesz) != 0) {
+		return -EINVAL;
+	}
+
+	mutexLock(strg->dev->ctx->lock);
+	while (tempsz < len) {
+		chunksz = len > strg->dev->mtd->oobSize ? strg->dev->mtd->oobSize : len;
+		memset(strg->dev->ctx->metabuf, 0xff, strg->dev->mtd->oobSize);
+		memcpy(strg->dev->ctx->metabuf, data + tempsz, chunksz);
+
+		res = flashdrv_write(strg->dev->ctx->dma, offs / strg->dev->mtd->writesz, NULL, strg->dev->ctx->metabuf);
+		if (res < 0) {
+			break;
+		}
+
+		tempsz += chunksz;
+		offs += strg->dev->mtd->writesz;
+	}
+	mutexUnlock(strg->dev->ctx->lock);
+
+	return res >= 0 ? len : res;
```

Comment:
Should be `return tempsz > 0 ? tempsz : res;` for consistent semantics with other functions.

Reply:
Right, it was left out.